gradle-failure-reports/src/main/java/com/palantir/gradle/failurereports/CompileFailuresService.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * (c) Copyright 2024 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.gradle.failurereports;
+
+import com.google.common.base.Splitter;
+import com.palantir.gradle.failurereports.CompileFailuresService.Parameters;
+import com.palantir.gradle.failurereports.junit.JunitReporter;
+import com.palantir.gradle.failurereports.util.FailureReporterResources;
+import java.io.File;
+import java.nio.file.Path;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.gradle.api.Project;
+import org.gradle.api.file.RegularFileProperty;
+import org.gradle.api.provider.Property;
+import org.gradle.api.provider.Provider;
+import org.gradle.api.services.BuildService;
+import org.gradle.api.services.BuildServiceParameters;
+import org.gradle.api.tasks.compile.AbstractCompile;
+
+public abstract class CompileFailuresService implements BuildService<Parameters>, AutoCloseable {
+
+    interface Parameters extends BuildServiceParameters {
+        RegularFileProperty getOutputFile();
+
+        RegularFileProperty getCompileOutputFile();
+
+        Property<File> getRootDir();
+    }
+
+    private static final Pattern COMPILE_ERROR_FIRST_LINE_PATTERN =
+            Pattern.compile("^(?<sourcePath>.*):(?<lineNumber>\\d+): (?<errorMessage>error: .*)$");
+    private static final Pattern COMPILE_ERROR_LAST_LINE_PATTERN = Pattern.compile("^\\d+ error(s)*$");
+    private static final Pattern COMPILE_ERROR_PATTERN =
+            Pattern.compile("(?m)^(?<sourcePath>.*):(?<lineNumber>\\d+): (?<errorMessage>error: (.|\\s)*)");
+
+    private final ConcurrentMap<String, Boolean> startedCompileErrorsByTaskPath = new ConcurrentHashMap<>();
+    private final ConcurrentMap<String, StringBuilder> compilerErrorsByTaskPath = new ConcurrentHashMap<>();
+
+    public final <T extends AbstractCompile> void maybeCollectErrorMessage(T task, CharSequence charSequence) {
```

Comment:
Does this char sequence always seem to be whole lines, or does it sometimes seem to be line fragments?

Reply:
They are always whole lines. I think the charSequence is received when the `endln` is encountered.